app/src/main/java/com/github/epfl/meili/MapActivity.kt

```diff
@@ -0,0 +1,145 @@
+        private val DEFAULT_ZOOM = 15
+        private val REQUEST_CODE: Int = 1
+    }
+
+    // API entry points
+    private lateinit var placesClient: PlacesClient
+    private lateinit var fusedLocationProviderClient: FusedLocationProviderClient
+
+    private lateinit var map: GoogleMap
+
+    private val defaultLocation = LatLng(48.864716, 2.349014) // Paris
+    private var location: Location? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_map)
+
+        // Initialize API entry points
+        Places.initialize(
+            applicationContext,
+            getString(R.string.google_maps_key)
+        ) // change API key here
+
+        this.placesClient = Places.createClient(this)
+        this.fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(this)
+
+        // Initialize map
+        val mapFragment = supportFragmentManager.findFragmentById(R.id.map) as SupportMapFragment?
+        mapFragment?.getMapAsync(this)
+    }
+
+    private fun isPermissionGranted(): Boolean {
+        return ContextCompat.checkSelfPermission(
+            this.applicationContext,
+            Manifest.permission.ACCESS_FINE_LOCATION
+        ) == PackageManager.PERMISSION_GRANTED
+    }
+
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<String>,
+        grantResults: IntArray
+    ) {
+//        // Check whether permission was granted
+//        if (requestCode == REQUEST_CODE && grantResults.isNotEmpty() &&
+//            grantResults[0] == PackageManager.PERMISSION_GRANTED
+//        )
+        updateMapUI()
+        getDeviceLocation()
+    }
+
+    override fun onMapReady(googleMap: GoogleMap) {
+        this.map = googleMap
+
+        val mapStyle = googleMap.setMapStyle(
+            MapStyleOptions.loadRawResourceStyle(
+                this, R.raw.map_style
+            )
+        )
+        if (!isPermissionGranted()) {
+            getLocationPermission()
+        }
+
+        updateMapUI()
+        getDeviceLocation()
+    }
+
+    private fun getLocationPermission() {
+        assert(!isPermissionGranted())
+        ActivityCompat.requestPermissions(
+            this, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
+            REQUEST_CODE
+        )
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun updateMapUI() {
+        if (isPermissionGranted()) {
+            map.isMyLocationEnabled = true
+            this.map.uiSettings?.isMyLocationButtonEnabled = true
+        } else {
+            map.isMyLocationEnabled = false
+            map.uiSettings?.isMyLocationButtonEnabled = false
+            location = null
+            Toast.makeText(this, "Cannot show location without permission", Toast.LENGTH_LONG)
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun getDeviceLocation() {
```

Comment:
Since I believe that we will be using the device location in a more parts of the project would it be possible that you created a Service that provides the location? What do you think?

Reply:
Yes, I can try to do that.

Reply:
In fact, the MapActivity is already using a LocationService provided by android. I tried abstracting away the permission handling in another class. However, I realized that it has to be done in whichever activity is using the location because it needs to be updated when the permission is granted.

Reply:
Couldn't we just ask for the permission at the very beginning after they login or when they start the app? Cause in the Sweng mini-project we managed to make a new service for the Location if I recall correctly, no?

Reply:
They were not using the fusedLocationProviderClient in the mini project. Here it does most of the job for us. As for the permissions, yes we could. But if we are only doing it once, I don't think we should factor it out. In the mini project last semester, the permission handling was done in the weather activity itself.

Reply:
I think we will still have to provide a LocationService but let's leave this as next steps. Thanks!